cardano-api-test/test/Test/Cardano/Api/TxBody.hs

```diff
@@ -0,0 +1,263 @@
+
+normalizeUpdateParameters
+  :: ProtocolParametersUpdate -> ProtocolParametersUpdate
+normalizeUpdateParameters params =
+  params
+    { protocolUpdateCollateralPercent =
+        zeroIsNothing $ protocolUpdateCollateralPercent params
+    , protocolUpdateMaxBlockExUnits =
+        zeroExecutionIsNothing $ protocolUpdateMaxBlockExUnits params
+    , protocolUpdateMaxCollateralInputs =
+        zeroIsNothing $ protocolUpdateMaxCollateralInputs params
+    , protocolUpdateMaxTxExUnits =
+        zeroExecutionIsNothing $ protocolUpdateMaxTxExUnits params
+    , protocolUpdateMaxValueSize =
+        zeroIsNothing $ protocolUpdateMaxValueSize params
+    , protocolUpdatePrices = zeroPricesIsNothing $ protocolUpdatePrices params
+    , protocolUpdateUTxOCostPerWord =
+        zeroIsNothing $ protocolUpdateUTxOCostPerWord params
+    }
+  where
+    zeroIsNothing :: (Eq a, Num a) => Maybe a -> Maybe a
+    zeroIsNothing = \case
+      Just 0 -> Nothing
+      x      -> x
+    zeroPricesIsNothing = \case
+      Just (ExecutionUnitPrices 0 0) -> Nothing
+      x                              -> x
+    zeroExecutionIsNothing = \case
+      Just (ExecutionUnits 0 0) -> Nothing
+      x                         -> x
+
+
+-- * Ugrading scripts
+--
+-- The instruction set from V1 may be used as V2,
+-- and we can't determine the language version from the transaction.
+-- When the user uses V1 instructions,
+-- we don't know if they used the V1 tag or the V2 tag.
+-- So it's safe to drop this information.
+
+upgradeScriptInEra :: ScriptInEra era -> ScriptInEra era
+upgradeScriptInEra = \case
+  ScriptInEra SimpleScriptV1InAllegra script ->
+    ScriptInEra SimpleScriptV2InAllegra $ upgradeScript script
+  ScriptInEra SimpleScriptV1InMary script ->
+    ScriptInEra SimpleScriptV2InMary $ upgradeScript script
+  ScriptInEra SimpleScriptV1InAlonzo script ->
+    ScriptInEra SimpleScriptV2InAlonzo $ upgradeScript script
+  other -> other
+
+upgradeScript :: Script SimpleScriptV1 -> Script SimpleScriptV2
+upgradeScript (SimpleScript SimpleScriptV1 script) =
+  SimpleScript SimpleScriptV2 $ upgradeSimpleScript script
+
+upgradeSimpleScript ::
+  SimpleScript SimpleScriptV1 -> SimpleScript SimpleScriptV2
+upgradeSimpleScript = \case
+  RequireSignature hash -> RequireSignature hash
+  RequireAllOf scripts  -> RequireAllOf $ map upgradeSimpleScript scripts
+  RequireAnyOf scripts  -> RequireAnyOf $ map upgradeSimpleScript scripts
+  RequireMOf n scripts  -> RequireMOf n $ map upgradeSimpleScript scripts
+
+
+review :: TxBodyContent BuildTx era -> TxBodyContent ViewTx era
+review TxBodyContent{..} =
+  TxBodyContent
+    { txCertificates    = reviewCertificates txCertificates
+    , txExtraScriptData = ViewTx
+    , txIns             = map reviewTxIn     txIns
+    , txMintValue       = reviewMintValue    txMintValue
+    , txProtocolParams  = ViewTx
+    , txWithdrawals     = reviewWithdrawals  txWithdrawals
+    , ..
```

Comment:
:+1: I think this is a legitimate use of record wildcards, so keep it, despite the typical style guide advice.

Reply:
Oops, I rewrote it without RWC before I found this comment. Should I return to RWC?